tests/project_helpers.py

```diff
@@ -0,0 +1,245 @@
+"""Common helpers shared between test_real_project and test_sample_projects."""
+import hashlib
+import logging
+import shlex
+import subprocess
+import sys
+from dataclasses import dataclass, fields
+from pathlib import Path
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Iterable,
+    Iterator,
+    List,
+    Optional,
+    Set,
+    Tuple,
+    Type,
+)
+
+import pytest
+
+from fawltydeps.main import Analysis
+from fawltydeps.types import TomlData
+
+if sys.version_info >= (3, 11):
+    import tomllib  # pylint: disable=E1101
+else:
+    import tomli as tomllib
+
+JsonData = Dict[str, Any]
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class CachedExperimentVenv:
+    """A virtualenv used in a sample or real-world experiment.
+
+    This is a test helper to create and reuse virtualenvs that contain real
+    packages used in our integration tests. The goal is that venvs are created
+    _once_ and reused as long as the commands used to create them and install
+    packages into them remain unchanged. These use pytest's caching
+    infrastructure meaning that the venvs typically end up living under
+    ~/.cache/pytest/d/...
+    """
+
+    requirements: List[str]  # PEP 508 requirements, passed to 'pip install'
+
+    def venv_script(self, venv_path: Path):
+        return (
+            [
+                f"rm -rf {venv_path}",
+                f"python3 -m venv {venv_path}",
+                f"{venv_path}/bin/pip install --upgrade pip",
+            ]
+            + [
+                f"{venv_path}/bin/pip install {shlex.quote(req)}"
+                for req in self.requirements
+            ]
+            + [
+                f"touch {venv_path}/.installed",
+            ]
+        )
+
+    def venv_hash(self):
+        """Returns a hash that depends on the venv script and python version.
+
+        The installation script will change if the code to setup the venv on
+        `venv_script` changes or if the requirements field of the experiment
+        changes. It will also be different for different Python versions.
+        The Python version currently used to run the tests is used to compute
+        the hash and create the venv.
+        """
+        dummy_script = self.venv_script(Path("/dev/null"))
+        py_version = f"{sys.version_info.major},{sys.version_info.major}"
+        script_and_version_bytes = ("".join(dummy_script) + py_version).encode()
```

Comment:
Hmm. It sort of works out, since we're only interested in calculating a hash over the contents here. That said I agree that it's not ideal.

A problem with changing this now is that it'll invalidate the current cache and have the next CI run be a lot slower.

@Nour-Mws is also currently working on setting up venvs as part of #252, and that code will likely look a bit different (less shell code), and I would consider rather rewriting this test code to reuse that functionality when it has been added. Will leave this as-is for now, I think.